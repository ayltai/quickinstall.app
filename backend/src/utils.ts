import { ErrorReporting, } from '@google-cloud/error-reporting';

const errors = new ErrorReporting();

export const handleError = (error : any) => {
    if (process.env.NODE_ENV === 'production') {
        errors.report(error);
    } else {
        console.error(error);
    }
};
