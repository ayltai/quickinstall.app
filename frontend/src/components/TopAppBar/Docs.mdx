import { Controls, Description, Meta, Title, Primary, Stories, } from '@storybook/blocks';

import * as ComponentStories from './TopAppBar.stories';

<Meta of={ComponentStories} />

<Title />

<Description />

# How to use a toolbar

A toolbar is a header component that should be presented at the top of a website. It can be customised by adding `children`.

# Default usage

<Primary />

# Parameters

<Controls sort='requiredFirst' />

<Stories
    title='Examples'
    includePrimary={false} />
