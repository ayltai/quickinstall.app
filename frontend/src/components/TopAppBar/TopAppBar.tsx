import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Container from '@mui/material/Container';
import Toolbar from '@mui/material/Toolbar';
import useScrollTrigger from '@mui/material/useScrollTrigger';
import React, { cloneElement, type FC, type ReactElement, } from 'react';
import { Link, } from 'react-router-dom';

import { useAppSelector, useDesktopMode, } from '../../hooks';
import type { TopAppBarProps, } from './TopAppBar.types';

const ElevationScroll = ({
    children,
} : {
    children : ReactElement,
}) => {
    const trigger = useScrollTrigger({
        disableHysteresis : true,
        threshold         : 0,
    });

    return cloneElement(children, {
        elevation : trigger ? 4 : 0,
    });
};

/**
 * A top app bar displays the title and actions related to the currently displayed content.
 * @param logo The logo to show at the start of the top app bar
 * @param title The title to show at the start of the top app bar and after the logo
 * @param actions The actions to show at the end of the top app bar
 * @param children The content to show at the middle of the top app bar
 * @param rest Other props
 */
export const TopAppBar : FC<TopAppBarProps> = ({
    logo,
    title,
    actions,
    children,
    ...rest
}) => {
    const { themeMode, } : { themeMode : 'light' | 'dark', } = useAppSelector(state => state.preference);

    const isDesktopMode = useDesktopMode();

    return (
        <>
            <ElevationScroll>
                <AppBar
                    style={{
                        backgroundColor      : themeMode === 'dark' ? 'rgba(0, 0, 0, 0.32)' : 'rgba(255, 255, 255, 0.32)',
                        backdropFilter       : 'blur(16px)',
                        WebkitBackdropFilter : 'blur(16px)',
                    }}
                    color='transparent'
                    {...rest}>
                    <Container
                        sx={{
                            paddingLeft : 0,
                        }}
                        maxWidth='lg'>
                        <Toolbar disableGutters>
                            <Link to='/'>
                                {logo}
                            </Link>
                            {isDesktopMode && title}
                            <Box flexGrow={1} />
                            {children}
                            <Box flexGrow={1} />
                            {actions}
                        </Toolbar>
                    </Container>
                </AppBar>
            </ElevationScroll>
            <Toolbar />
        </>
    );
};
