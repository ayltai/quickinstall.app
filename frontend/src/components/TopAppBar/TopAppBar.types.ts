import type { AppBarOwnProps, } from '@mui/material/AppBar';
import type { ReactNode, } from 'react';

export interface TopAppBarProps extends AppBarOwnProps {
    logo?     : ReactNode,
    title?    : ReactNode,
    actions?  : ReactNode,
    children? : ReactNode,
}
