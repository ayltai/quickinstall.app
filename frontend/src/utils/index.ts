import * as Sentry from '@sentry/react';
import { FirebaseApp, } from 'firebase/app';
import { getAnalytics, logEvent, } from 'firebase/analytics';

export const handleError = (error : any) => {
    if (import.meta.env.PROD) {
        Sentry.captureException(error);
    } else {
        console.error(error);
    }
};

export const logPageView = (app : FirebaseApp, {
    pageName,
    route,
} : {
    pageName : string,
    route    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'page_view', {
            page_title    : pageName,
            page_location : route,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logAddToCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'add_to_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logRemoveFromCart = (app : FirebaseApp, {
    platform,
    itemId,
} : {
    platform : 'win' | 'mac',
    itemId   : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'remove_from_cart', {
            items : [
                {
                    item_id : `${platform}/${itemId}`,
                },
            ],
        });
    } catch (error) {
        handleError(error);
    }
};

export const logCheckout = (app : FirebaseApp, {
    platform,
    itemIds,
} : {
    platform : 'win' | 'mac',
    itemIds  : string[],
}) => {
    try {
        logEvent(getAnalytics(app), 'checkout', {
            items : itemIds.map((itemId : string) => ({
                item_id : `${platform}/${itemId}`,
            })),
        });
    } catch (error) {
        handleError(error);
    }
};

export const logSearch = (app : FirebaseApp, {
    platform,
    query,
} : {
    platform : 'win' | 'mac',
    query    : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'search', {
            search_term : `${platform}/${query}`,
        });
    } catch (error) {
        handleError(error);
    }
};

export const logShare = (app : FirebaseApp, {
    product,
    itemId,
} : {
    product : 'win/packages' | 'mac/packages' | 'tweaks',
    itemId  : string,
}) => {
    try {
        logEvent(getAnalytics(app), 'share', {
            content_type : product,
            item_id      : itemId,
        });
    } catch (error) {
        handleError(error);
    }
};
