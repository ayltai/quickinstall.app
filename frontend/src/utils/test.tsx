import { createTheme, ThemeProvider, } from '@mui/material';
import { configureStore, type Store, } from '@reduxjs/toolkit';
import { render, } from '@testing-library/react';
import React, { type ReactNode, } from 'react';
import { Provider, } from 'react-redux';
import { MemoryRouter, } from 'react-router-dom';

import { appStoreApi, chocolateyApi, contentfulApi, homebrewApi, quickInstallApi, } from '../apis';
import { recipeReducer, preferenceReducer, spellReducer, } from '../states';
import { appTheme, } from '../styles';

export const createStore = (preloadedState? : unknown) => configureStore({
    preloadedState,
    reducer    : {
        recipe                          : recipeReducer,
        preference                      : preferenceReducer,
        spell                           : spellReducer,
        [ appStoreApi.reducerPath  ]    : appStoreApi.reducer,
        [ chocolateyApi.reducerPath ]   : chocolateyApi.reducer,
        [ contentfulApi.reducerPath ]   : contentfulApi.reducer,
        [ homebrewApi.reducerPath  ]    : homebrewApi.reducer,
        [ quickInstallApi.reducerPath ] : quickInstallApi.reducer,
    },
    middleware : getDefaultMiddleware => getDefaultMiddleware().concat(appStoreApi.middleware, chocolateyApi.middleware, contentfulApi.middleware, homebrewApi.middleware, quickInstallApi.middleware),
});

export const defaultStore = createStore();

export type RootState = ReturnType<typeof defaultStore.getState>;

// @ts-ignore
const theme = createTheme(appTheme);

const customRender = (ui : any, {
    preloadedState,
    store = createStore(preloadedState),
    ...rest
} : {
    preloadedState?   : Record<string, any>,
    store?            : Store<RootState>,
    [ rest : string ] : any,
} = {}) => render(ui, {
    wrapper : ({
        children,
    } : {
        children : ReactNode,
    }) => (
        <Provider store={store}>
            <MemoryRouter>
                <ThemeProvider theme={theme}>
                    {children}
                </ThemeProvider>
            </MemoryRouter>
        </Provider>
    ),
    ...rest,
});

export * from '@testing-library/react';

export { customRender as render, };
