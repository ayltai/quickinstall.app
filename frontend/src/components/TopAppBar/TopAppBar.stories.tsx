import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import type { Meta, StoryObj, } from '@storybook/react';
import React from 'react';

import { DarkModeSwitch, } from '../DarkModeSwitch';
import { ProductSelection, } from '../ProductSelection';
import { ShoppingCart, } from '../ShoppingCart';
import SmallLogo from './SmallLogo.webp';
import { TopAppBar, } from './TopAppBar';

const meta : Meta<typeof TopAppBar> = {
    component : TopAppBar,
    title     : 'Components/Data Presentations/Toolbar',
    argTypes  : {
        logo : {
            description : 'The logo to show at the start of the top app bar',
        },
        title : {
            description : 'The title to show at the start of the top app bar and after the logo',
        },
        actions : {
            description : 'The actions to show at the end of the top app bar',
        },
        children : {
            description : 'The content to show at the middle of the top app bar',
        },
    },
};

export default meta;

type Story = StoryObj<typeof TopAppBar>;

export const Default : Story = {
    args : {
        logo     : (
            <img
                style={{
                    marginRight : 8,
                }}
                src={SmallLogo}
                alt='Logo'
                width={48}
                height={48} />
        ),
        title    : (
            <Typography
                variant='h4'
                fontWeight='bold'>
                QuickInstall.app
            </Typography>
        ),
        actions  : (
            <>
                <DarkModeSwitch />
                <Box padding={1} />
                <ShoppingCart />
            </>
        ),
        children : (
            <ProductSelection
                sx={{
                    marginLeft : -16,
                }} />
        ),
    },
    parameters : {
        layout : 'padded',
    },
};
