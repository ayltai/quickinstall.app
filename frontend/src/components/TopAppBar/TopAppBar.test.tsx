import React from 'react';

import { render, } from '../../utils/test';
import { TopAppBar, } from './TopAppBar';

describe('<TopAppBar />', () => {
    it('renders correctly', () => expect(render(
        <TopAppBar
            logo='Logo'
            title='Title'
            actions='Actions'>
            Children
        </TopAppBar>
    )).toMatchSnapshot());
});
