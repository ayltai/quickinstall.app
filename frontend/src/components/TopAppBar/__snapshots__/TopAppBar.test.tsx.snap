// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TopAppBar /> > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionFixed mui-fixed css-1jvst5f-MuiPaper-root-MuiAppBar-root"
        style="background-color: rgba(255, 255, 255, 0.32);"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-kicy4c-MuiContainer-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
          >
            <a
              href="/"
            >
              Logo
            </a>
            <div
              class="MuiBox-root css-i9gxme"
            />
            Children
            <div
              class="MuiBox-root css-i9gxme"
            />
            Actions
          </div>
        </div>
      </header>
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      />
    </div>
  </body>,
  "container": <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionFixed mui-fixed css-1jvst5f-MuiPaper-root-MuiAppBar-root"
      style="background-color: rgba(255, 255, 255, 0.32);"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-kicy4c-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
        >
          <a
            href="/"
          >
            Logo
          </a>
          <div
            class="MuiBox-root css-i9gxme"
          />
          Children
          <div
            class="MuiBox-root css-i9gxme"
          />
          Actions
        </div>
      </div>
    </header>
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
